app/Services/AuthenticationService.php

```diff
@@ -4,44 +4,75 @@
+use Illuminate\Contracts\Hashing\Hasher;
+use Illuminate\Database\Eloquent\ModelNotFoundException;
 use Illuminate\Http\Exceptions\HttpResponseException;
 use Illuminate\Http\Request;
-use Illuminate\Support\Facades\Hash;
-use Laravel\Socialite\Contracts\User as FacebookUser;
+use Illuminate\Http\Response;
+use Illuminate\Support\Facades\Schema;
+use Laravel\Socialite\Contracts\User as SocialUser;
 
 class AuthenticationService
 {
+    protected Hasher $hash;
+    public function __construct(Hasher $hash)
+    {
+        $this->hash = $hash;
+    }
+
     /**
      * @throws HttpResponseException
      */
     public function login(Request $request): string
     {
         $user = User::where("email", $request->email)->first();
-
-        if ($user === null || !Hash::check($request->password, $user->password)) {
-            throw new HttpResponseException(response()->json(["message" => __("auth.credentialsMismatch")], 401));
+        if ($user === null || !$this->hash->check($request->password, $user->password)) {
+            throw new HttpResponseException(response()->json(["message" => __("auth.credentialsMismatch")], Response::HTTP_UNAUTHORIZED));
         }
         return $user->createToken($user->email)->plainTextToken;
     }
 
     public function register(array $input): void
     {
-        $input["password"] = Hash::make($input["password"]);
+        $input["password"] = $this->hash->make($input["password"]);
         $user = User::create($input);
         $user->save();
     }
 
-    public function authorizeByFacebook(FacebookUser $facebookUser): string
+    /**
+     * @throws HttpResponseException
+     */
+    public function getTokenBySocialLogin(SocialUser $socialUser, string $socialProviderName): string
     {
-        $user = User::where("facebook_id", $facebookUser->getId())->first();
-        if (!$user) {
+        if (!$this->isSocialProviderConfigured($socialProviderName)) {
+            throw new HttpResponseException(response()->json(["message" => __("auth.socialServiceError")], Response::HTTP_NOT_IMPLEMENTED));
+        }
+        try {
+            $user = SocialProfile::query()->where($socialProviderName . "_id", $socialUser->getId())->firstOrFail()->user;
+        } catch (ModelNotFoundException $ex) {
             $user = new User();
-            $user->name = $facebookUser->getName();
-            $user->email = $facebookUser->getEmail();
-            $user->facebookId = $facebookUser->getId();
+            $user->name = $socialUser->getName();
+            $user->email = $socialUser->getEmail();
             $user->save();
+
+            $socialProfile = new SocialProfile();
+            $socialProfile->userId = $user->id;
+            $socialProfile->facebookId = $socialUser->getId();
+            $socialProfile->save();
         }
+
         return $user->createToken($user->email)->plainTextToken;
     }
+
+    private function isSocialProviderConfigured(string $socialProviderName): bool
```

Comment:
Okaaay, doing that magic stuff with every user would be an overkill, I think. Maybe we could just put registered social providers in some config file, so it could be easily cacheable and retrievable?

Reply:
Yh, okay, I can see now that you're using config. So maybe this `isSocialProviderConfigured` should be some kind of test in CI?